IPython/core/history.py

```diff
@@ -81,18 +80,29 @@ def catch_corrupt_db(f, self, *a, **kw):
     try:
         return f(self, *a, **kw)
     except (DatabaseError, OperationalError) as e:
-        if os.path.isfile(self.hist_file):
-            # Try to move the file out of the way
-            base,ext = os.path.splitext(self.hist_file)
-            newpath = base + '-corrupt' + ext
-            os.rename(self.hist_file, newpath)
-            print("ERROR! History file wasn't a valid SQLite database (%s)." % e,
-            "It was moved to %s" % newpath, "and a new file created.", file=sys.stderr)
+        self._corrupt_db_counter += 1
+        self.log.error("Failed to open SQLite history %s (%s).", self.hist_file, e)
+        if self.hist_file != ':memory:':
+            if self._corrupt_db_counter > self._corrupt_db_limit:
+                self.hist_file = ':memory:'
+                self.log.error("Failed to load history too many times, history will not be saved.")
+            elif os.path.isfile(self.hist_file):
+                # Try to move the file out of the way
+                base,ext = os.path.splitext(self.hist_file)
+                now = datetime.datetime.now().isoformat().replace(':', '.')
+                newpath = base + '-corrupt-' + now + ext
```

Comment:
That would mean starting IPython twice wipes out your history. In the NFS case prompting this, the db is not actually bad, SQLite is just failing to open it due to locking problems. So we should be making sure that we never overwrite the 'good' db, otherwise the '-corrupt' backup is pretty pointless.